src/internal/compiler/template_instantiation.rs

```diff
@@ -318,6 +330,116 @@ fn instantiate_choice(
     new_type_ref
 }
 
+fn instantiate_union(
+    pkg: &mut ZPackage,
+    scope: &mut ModelScope,
+    z_union: &ZUnion,
+    template_arguments: &[TypeReference],
+    type_arguments: &[Rc<RefCell<Expression>>],
+    instantiated_name: &String,
+) -> TypeReference {
+    assert!(!z_union.template_parameters.is_empty());
+    assert!(z_union.template_parameters.len() == template_arguments.len());
+
+    let new_type_ref = TypeReference {
+        is_builtin: false,
+        package: pkg.name.clone(),
+        name: instantiated_name.clone(),
+        bits: 0,
+        template_arguments: vec![],
+        type_arguments: type_arguments.to_owned(),
+        length_expression: None,
+    };
+
+    if pkg.zunions.contains_key(instantiated_name) {
+        // The union was already instantiated, there is no need to instantiate it again.
+        // Return a new reference to the already existing type.
+        return new_type_ref;
+    }
```

Comment:
I'm wondering if it is possible to use a single TypeReference instance? Creating a new instance with copies of three strings every time does not feel optimal.

The code also seems to rely on `instantiated_name` being descriptive enough to include difference in type arguments, without documenting that assumption. Is this guaranteed by all callers?

Reply:
Yes, I am also not happy with the many copy operations. I think the template instantiation requires a bit of refactoring, but let me work on that in a different PR.